### PR TITLE
tablecodec,mysql, util: mysql TypeDuration Fsp 

### PR DESCRIPTION
**Pull request by dreamquster**
Fixed RoundError #3471
Bug Cause:
1. when deserialize mysqlDuration, the Fsp is setted with 0 instead of the Fsp from column information.
2. the default decimal for column with TypeDuration is 0.
3. 'show create table' statement print  time instead of time(0) when the fsp of TypeDuration is 0.

### CODE REVIEW

File: util/types/field_type.go [hunk 1]
```diff
@@ -127,7 +127,7 @@ func (ft *FieldType) CompactStr() string {
 			es = append(es, e)
 		}
 		suffix = fmt.Sprintf("('%s')", strings.Join(es, "','"))
-	case mysql.TypeTimestamp, mysql.TypeDatetime, mysql.TypeDate:
+	case mysql.TypeTimestamp, mysql.TypeDatetime, mysql.TypeDate, mysql.TypeDuration:
```
Comment (shenli) [inline util/types/field_type.go]: Could you add an unit test for this?

File: tablecodec/tablecodec.go [hunk 2]
```diff
@@ -421,8 +421,8 @@ func unflatten(datum types.Datum, ft *types.FieldType, loc *time.Location) (type
 		}
 		datum.SetMysqlTime(t)
 		return datum, nil
-	case mysql.TypeDuration:
-		dur := types.Duration{Duration: time.Duration(datum.GetInt64())}
+	case mysql.TypeDuration: //duration should read fsp from column meta data
```
Comment (shenli) [inline tablecodec/tablecodec.go]: Please add an unit test in tablecode_test.go

Comment (dreamquster) [inline tablecodec/tablecodec.go]: add c4 cases in tablecode_test.go

File: util/types/field_type.go (hunk 1 above)
Comment (dreamquster) [inline util/types/field_type.go]: In testSuite.TestShow, the following code verify the case that it should print time when fsp of time is 0 or -1.
`testSQL = `create table ptest(
		a int primary key,
		b double NOT NULL DEFAULT 2.0,
		c varchar(10) NOT NULL,
		d time unique,
		e timestamp NULL
	);`
	tk.MustExec(testSQL)
	testSQL = "show create table ptest;"
	result = tk.MustQuery(testSQL)
	c.Check(result.Rows(), HasLen, 1)
	row = result.Rows()[0]
	expectedRow = []interface{}{
		"ptest", "CREATE TABLE `ptest` (\n  `a` int(11) NOT NULL,\n  `b` double NOT NULL DEFAULT '2.0',\n  `c` varchar(10) NOT NULL,\n  `d` time DEFAULT NULL,\n  `e` timestamp NULL DEFAULT NULL,\n PRIMARY KEY (`a`),\n  UNIQUE KEY `d` (`d`)\n) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin"}
	for i, r := range row {
		c.Check(r, Equals, expectedRow[i])
	}`

File: tablecodec/tablecodec.go (hunk 2 above)
Comment (dreamquster) [inline tablecodec/tablecodec.go]: why the status of travis-ci  is queued after hours?

File: util/types/field_type.go (hunk 1 above)
Comment (shenli) [inline util/types/field_type.go]: So why the test case does not failed without this change?

File: tablecodec/tablecodec.go (hunk 2 above)
Comment (shenli) [inline tablecodec/tablecodec.go]: ok

File: util/types/field_type.go (hunk 1 above)
Comment (dreamquster) [inline util/types/field_type.go]: Without this change,  'd time unique' create a duration field that it's Fsp is -1(UnspecifiedLength).
 `ft.Decimal != UnspecifiedLength`  prevent 'show create table x' print length of field. 
Applied this PR, the default fsp of duration is 0 when length of duration field is not set。So, duration should be printed length only with length > 0 as other time types do.

**Comment by dreamquster** [comment]
@XuHuaiyu , Add test case in execute_test.go.

File: executor/executor_test.go [hunk 3]
```diff
@@ -1615,3 +1615,17 @@ func (s *testSuite) TestSelectForUpdate(c *C) {
 
 	tk1.MustExec("commit")
 }
+
+func (s *testSuite) TestCastTimeAsSig(c *C) {
+	defer func() {
+		s.cleanEnv(c)
+		testleak.AfterTest(c)()
+	}()
+	// Issue#3471
```
Comment (XuHuaiyu) [inline executor/executor_test.go]: put this test case in `TestBuiltin()` in this file.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**